src/fragmentarium/ui/display/Transliteration.tsx

```diff
@@ -109,32 +112,89 @@ class LineAccumulator {
     )
   }
 }
+function DisplayLineNumberRange({ start, end }: LineNumberRange): JSX.Element {
+  const startPrime = start.hasPrime ? '′' : ''
+  const endPrime = end.hasPrime ? '′' : ''
+  const startPrefixMod = start.prefixModifier ? start.prefixModifier + '+' : ''
+  const startSuffixMod = start.suffixModifier ? start.suffixModifier : ''
+  const endPrefixMod = end.prefixModifier ? end.prefixModifier + '+' : ''
+  const endSuffixMod = end.suffixModifier ? end.suffixModifier : ''
+  const result = `(${startPrefixMod}${start.number}${startPrime}${startSuffixMod}-${endPrefixMod}${end.number}${endPrime}${endSuffixMod})`
+  return (
+    <span>
+      <sup>{result}</sup>
+    </span>
+  )
+}
 
+function DisplayLineNumber({ hasPrime, number }: LineNumber): JSX.Element {
+  const prime = hasPrime ? '′' : ''
+  return (
+    <span>
+      <sup>{`(${number}${prime})`}</sup>
+    </span>
+  )
+}
+const lineNumberTypeToComponent = {
+  LineNumber: DisplayLineNumber,
+  LineNumberRange: DisplayLineNumberRange
+}
+
+function DisplayTextLine({
+  line,
+  container = 'div'
+}: {
+  line: Line
+  container?: string
+}): JSX.Element {
+  const textLine = line as TextLine
+  const lineTypeComponent =
+    lineNumberTypeToComponent[textLine.lineNumber.type as string]
+  return React.createElement(
+    container,
+    { className: classNames([`Transliteration__${textLine.type}`]) },
+    [
+      React.createElement(lineTypeComponent, {
+        key: 0,
+        ...textLine.lineNumber
+      }),
+      ...helperDisplayLineTokens(line.content)
+    ]
+  )
+}
 function DisplayLine({
-  line: { type, prefix, content },
+  line,
   container = 'div'
 }: {
   line: Line
   container?: string
 }): JSX.Element {
   return React.createElement(
     container,
-    { className: classNames([`Transliteration__${type}`]) },
+    { className: classNames([`Transliteration__${line.type}`]) },
     [
-      <span key="prefix">{prefix}</span>,
-      ...content.reduce((acc: LineAccumulator, token: Token) => {
-        if (isShift(token)) {
-          acc.applyLanguage(token)
-        } else if (isDocumentOrientedGloss(token)) {
-          token.side === 'LEFT' ? acc.openGloss() : acc.closeGloss()
-        } else {
-          acc.pushToken(token)
-        }
-        return acc
-      }, new LineAccumulator()).result
+      <span key="prefix">{line.prefix}</span>,
+      ...helperDisplayLineTokens(line.content)
     ]
```

Comment:
There is no need to have an array here, you can just have two elements: the prefix and the content.